### ISSUE TITLE
metrics in multi-gpu settings

### ISSUE THREAD

**Issue by juliusberner** [PyTorchLightning/pytorch-lightning#681]
1. Question regarding the logs in ddp mode: 
Logs (in `metrics.csv`) only seem to contain metrics of the model/batches on the gpu with lowest index, although they are computed for each gpu according to `validation_end`.

2. (Minor) issue regarding the dp/ddp2 mode:
 The documentation suggests averaging the metrics as
`if self.trainer.use_dp or self.trainer.use_ddp2:`
   `metric_value = torch.mean(metric_value)`
However, this might be (slightly) wrong if the batchsize is not divisible by the number of gpus.

**Comment by williamFalcon** [comment]
1. yes.
2. in validation_end you are given the output of each batch, so a mean there would be appropriate.

does that answer your qs?

**Comment by juliusberner** [comment]
Thank you for your prompt reply. 

1. I think it would be desirable to log metrics for each gpu in order to be able to compute the metrics over the whole dataset afterwards (as each gpu only sees a fraction of the dataset)

2. Take, for instance, batch_size=39, distributed-backend=dp, and gpus=2. 
Then, gpu_0 and gpu_1 receive batches of size 20 and 19, respectively.  
If we compute, e.g., the accuracy for the whole batch the correct formula would be of the form
`metric_value[0] * 20/39 + metric_value[1] * 19/39`
(this 'problem' occurs whenever the batchsize is not divisible by the number of gpus)
